README.org
[[http://www.gnu.org/licenses/gpl-3.0.txt][file:https://img.shields.io/badge/license-GPL_3-orange.svg]]
[[http://melpa.org/#/bui][file:http://melpa.org/packages/bui-badge.svg]]
[[http://stable.melpa.org/#/bui][file:http://stable.melpa.org/packages/bui-badge.svg]]

* About

=BUI= (=Buffer User Interface=) is an Emacs library that can be used to
make user interfaces to display some kind of entries (like packages,
buffers, functions, etc.).

The intention of BUI is to be a high-level library which is convenient
to be used both by:

- *package makers*, as there is no need to bother about implementing
  routine details and usual features (like buffer history, filtering
  displayed entries, etc.);

- *users*, as it provides familiar and intuitive interfaces with usual
  keys (for moving by lines, marking, sorting, switching between
  buttons); and what is also important, the defined interfaces are
  highly configurable through various generated variables.  A summary of
  available key bindings can be displayed by pressing =h=.

* Usage

BUI provides means to display entries in 2 types of buffers:

- =list=: it is based on =tabulated-list-mode=, thus it looks similar to
  a list of Emacs packages (=M-x list-packages=);

- =info=: it can be used to display more verbose info, like various
  buttons, text and other stuff related to the displayed entry (or
  entries).

In short, you define how a =list= / =info= interface looks like (using
=bui-define-interface= macro), and then you can make some user commands
that will display entries (using =bui-get-display-entries= and similar
functions).

For example, you can make a =list= interface to display buffers (similar
to what =M-x list-buffers= do), like this:

#+BEGIN_SRC emacs-lisp
(require 'bui)

(defun buffers-buffer->entry (buffer)
  (with-current-buffer buffer
    `((id   . ,buffer)
      (name . ,(buffer-name))
      (mode . ,major-mode)
      (size . ,(buffer-size))
      (file-name . ,buffer-file-name))))

(defun buffers-get-entries ()
  (mapcar 'buffers-buffer->entry (buffer-list)))

(bui-define-interface buffers list
  :buffer-name "*Buffers*"
  :get-entries-function 'buffers-get-entries
  :format '((name nil 30 t)
            (mode nil 25 t)
            (size nil 8 bui-list-sort-numerically-2 :right-align t)
            (file-name bui-list-get-file-name 30 t))
  :sort-key '(name))

(defun buffers ()
  "Display a list of buffers."
  (interactive)
  (bui-get-display-entries 'buffers 'list))
#+END_SRC

This is a simplified example just to demonstrate how =bui.el= can be
used.  For full example see [[file:examples/buffers.el]].  You can see how
it looks like on the following screenshot.  =M-x buffers= displays a
list of buffers, then 2 buffers are marked (with =m= key) and
"described" in =info= buffer (with =i= key).

[[http://i.imgur.com/3dlBu2Y.png]]

=bui-define-interface= macro takes the following arguments:

- ENTRY-TYPE: an arbitrary symbol to denote the entry type.

- BUFFER-TYPE: =list= or =info= symbol.

- Keyword arguments: used to define various interface parameters and to
  set default values of user variables (like titles, buffer name,
  etc.).  The main keywords that should be specified are:

  + =:get-entries-function=: this function should return a list of
    entries to display.  Each entry is a usual association list with one
    required =id= key (it is used to fill =tabulated-list-entries=
    variable).

  + =:format=: it specifies how the data is displayed; see docstrings of
    the generated =ENTRY-TYPE-BUFFER-TYPE-format= variables for details
    (in the current example: =buffers-list-format= and
    =buffers-info-format=).

For real-world examples you may look at [[https://github.com/alezost/aurel][aurel]] or [[https://github.com/alezost/guix.el][guix]] packages.


NEWS
-*- mode: org -*-

* Changes in 1.2.1 (since 1.2.0)

The only change is: if list format function is specified (in
'ENTRY-TYPE-list-format' variable), it is called even if an entry does
not have an according parameter.  This allows to use non-existing
parameters in the format variable, for example, if this variable has
(fake some-fun 10 t) specification in it, 'some-fun' will be called even
if 'fake' parameter does not exist.  Previously, the function was
ignored in such cases, and 'bui-empty-string' was used instead.

* Changes in 1.2.0 (since 1.1.0)

** New functions

- bui-list-marked-or-current
- bui-list-map-marked

** New variables

- bui-hint-format

** Minor performance improvements

* Changes in 1.1.0 (since 1.0.1)

** Hint system is added

A user can press "h" (or any unbound key) in any 'list' or 'info' buffer
to display a "hint" (a message with available key bindings).  Also
'bui-define-interface' macro supports ':hint' argument (thus
'ENTRY-TYPE-BUFFER-TYPE-hint' variable is generated) to override the
default hint.

** Changes in mode initializing

Previously, a function specified at ':mode-init-function' interface
keyword should have called 'bui-mode-initialize-default'.  It is not
needed anymore ('bui-mode-initialize-default' is deprecated and
ignored).

** New functions

- 'bui-current-param-title'

** Fixed bugs

*** History size variable is honored

Previously, setting 'ENTRY-TYPE-BUFFER-TYPE-history-size' did not take
effect.

*** 'bui-list-show-single' variable is honored

Previously, if 'ENTRY-TYPE-BUFFER-TYPE-show-single' was set to nil and
'bui-list-show-single' to t, a single entry was displayed in 'info'
buffer anyway; now it is displayed in 'list' buffer.

* Changes in 1.0.1 (since 1.0)

** Several minor bugs fixed

** Only (require 'bui) is needed

Previously, a package maker had to require various BUI features
depending on what bui code is used; now only (require 'bui) is needed as
it loads all the rest BUI features.

* 1.0 (initial release)

** History

The ancestor of BUI was some code written in February 2014 as a part of
[[https://github.com/alezost/aurel][aurel]] package.  Then this code was copied to [[https://github.com/alezost/guix.el][guix.el]], and was heavily
developed for about 2 years as a part of the Emacs interface for Guix.
Finally, in September 2016, this code was extracted from "guix.el" and
became "bui" library (later, both "aurel" and "Emacs-Guix" were switched
to BUI).


examples/buffers.el
;;; buffers.el --- List of buffers and buffer info

;; Copyright © 2016–2017 Alex Kost <alezost@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; This is an example of using BUI (Buffer User Interface) library.
;;
;; It shows how to use bui to display a list of buffers à la
;; `list-buffers' or `ibuffer'.
;;
;; To try it, load this file (for example, with "M-x load-file"), and
;; run "M-x buffers" command.  There you can mark several buffers (with
;; "m") and press "i" to display the info buffer; press "f f" to enable
;; fiters, etc.  Press "h" to look at the "hint" (available keys).

;;; Code:

(require 'bui)
(require 'help-mode)  ; for `help-function' button

(bui-define-groups buffers
  :parent-group tools
  :parent-faces-group faces
  :group-doc "Settings for '\\[buffers]' command."
  :faces-group-doc "Faces for '\\[buffers]' command.")

(defun buffers-get-buffers (&optional search-type &rest search-values)
  (or search-type (setq search-type 'all))
  (cl-case search-type
    (all (buffer-list))
    (id search-values)
    (t (error "Unknown search type: %S" search-type))))

(defun buffers-buffer-file-name ()
  (or buffer-file-name
      (and (boundp 'dired-directory)
           (if (stringp dired-directory)
               dired-directory
             (car dired-directory)))))

(defun buffers-visited-file-modtime ()
  (let ((time (visited-file-modtime)))
    (cl-case time
      ((-1 0) nil)
      (t time))))

(defun buffers-buffer->entry (buffer)
  (with-current-buffer buffer
    `((id   . ,buffer)
      (name . ,(buffer-name))
      (mode . ,major-mode)
      (size . ,(buffer-size))
      (mod-time  . ,(buffers-visited-file-modtime))
      (file-name . ,(buffers-buffer-file-name)))))

(defun buffers-get-entries (&rest args)
  (mapcar #'buffers-buffer->entry
          (apply #'buffers-get-buffers args)))

(bui-define-entry-type buffers
  :titles '((mod-time . "Modification Time"))
  :get-entries-function #'buffers-get-entries
  :filter-predicates
  '(buffers-buffer-ephemeral?
    buffers-buffer-non-ephemeral?
    buffers-buffer-visiting-file?
    buffers-buffer-not-visiting-file?))

(defun buffers-describe-mode-function (button)
  (describe-function (intern (button-label button))))


;;; Filter predicates

(defun buffers-buffer-ephemeral? (entry)
  "Return non-nil, if ENTRY's buffer name starts with a space."
  (string= " " (substring (bui-entry-value entry 'name) 0 1)))

(defun buffers-buffer-non-ephemeral? (entry)
  "Return non-nil, if ENTRY's buffer name does not start with a space."
  (not (buffers-buffer-ephemeral? entry)))

(defun buffers-buffer-visiting-file? (entry)
  "Return non-nil, if ENTRY's buffer visits a file."
  (bui-entry-non-void-value entry 'file-name))

(defun buffers-buffer-not-visiting-file? (entry)
  "Return non-nil, if ENTRY's buffer does not visit a file."
  (not (buffers-buffer-visiting-file? entry)))


;;; 'Info' interface

(bui-define-interface buffers info
  :format '((name format buffers-info-insert-name)
            (mode format (simple buffers-mode-function))
            (size format (format))
            nil
            (file-name nil (simple bui-file))
            (mod-time format (time))))

(define-button-type 'buffers-mode-function
  :supertype 'help-function
  'action 'buffers-describe-mode-function)

(defun buffers-info-insert-name (name entry)
  (bui-info-insert-value-simple (bui-entry-value entry 'name)
                                'mode-line-buffer-id)
  (bui-insert-indent)
  (bui-insert-action-button
   "Switch"
   (lambda (btn)
     (pop-to-buffer (button-get btn 'buffer)))
   "Switch to this buffer"
   'buffer (bui-entry-id entry)))


;;; 'List' interface

(bui-define-interface buffers list
  :buffer-name "*Buffers*"
  :describe-function #'buffers-list-describe
  :titles '((mod-time . "Mod. Time"))
  :format '((name nil 30 t)
            (mode buffers-list-get-mode 25 t)
            (size nil 8 bui-list-sort-numerically-2 :right-align t)
            ;; (mod-time bui-list-get-time 20 t)
            (file-name bui-list-get-file-name 30 t))
  :hint 'buffers-list-hint
  :sort-key '(name))

(let ((map buffers-list-mode-map))
  (define-key map (kbd "RET") 'buffers-list-switch-to-buffer)
  (define-key map (kbd "k")   'buffers-list-kill-buffers))

(defvar buffers-list-default-hint
  '(("\\[buffers-list-switch-to-buffer]") " switch to buffer;\n"
    ("\\[buffers-list-kill-buffers]") " kill buffer(s);\n"))

(defun buffers-list-hint ()
  "Return a hint string to display in the echo area."
  (bui-format-hints
   buffers-list-default-hint
   (bui-default-hint)))

(defun buffers-list-get-mode (mode &optional _)
  "Return MODE button specification for `tabulated-list-entries'.
MODE may be nil."
  (list (symbol-name mode)
        :supertype 'help-function
        'action 'buffers-describe-mode-function))

(defun buffers-list-describe (&rest buffers)
  "Display 'info' buffer for BUFFERS."
  (bui-get-display-entries 'buffers 'info (cons 'id buffers)))

(defun buffers-list-switch-to-buffer ()
  (interactive)
  (pop-to-buffer (bui-list-current-id)))

(defun buffers-list-kill-buffers ()
  "Kill marked buffers (or the current buffer)."
  (interactive)
  (dolist (buffer (or (bui-list-get-marked-id-list)
                      (list (bui-list-current-id))))
    (kill-buffer buffer))
  (revert-buffer nil t))


;;; Interactive commands

;;;###autoload
(defun buffers ()
  "Display a list of buffers."
  (interactive)
  (bui-list-get-display-entries 'buffers))

(provide 'buffers)

;;; buffers.el ends here
